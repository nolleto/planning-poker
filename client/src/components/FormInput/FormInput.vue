<template>
  <div>
    <input
      :class="[
        bootstrap.formControl,
        {
          [bootstrap.isValid]: state === 'valid',
          [bootstrap.isInvalid]: state === 'invalid'
        }
      ]"
      :name="name"
      :id="name"
      :value="value"
      v-bind="$attrs"
      v-on="listeners"
    />

    <form-input-feedback :state="state">
      {{ error }}
    </form-input-feedback>
  </div>
</template>

<script>
import styles from '@/styles'

export default {
  name: 'FormInput',

  $_veeValidate: {
    value () {
      return this.$el.value
    },
    name () {
      return this.name
    }
  },

  inheritAttrs: false,

  props: {
    name: {
      type: String,
      required: true
    },

    value: {
      type: String
    },

    state: {
      type: String
    },

    error: {
      type: String
    }
  },

  computed: {
    bootstrap: () => styles.bootstrap,

    listeners () {
      return {
        ...this.$listeners,
        input: e => this.$emit('input', e.target.value)
      }
    }
  },

  mounted () {
    this.$el.value = this.value
  }
}
</script>
