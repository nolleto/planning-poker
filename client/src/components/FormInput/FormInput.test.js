import FormInput from './index'
import { shallowMount } from '@vue/test-utils'

const defaultProps = {
  name: 'username'
}

const shallowComponent = (opts = { propsData: defaultProps }) => {
  return shallowMount(FormInput, opts)
}

describe('FormInput component', () => {
  let wrapper

  describe('Lifecycle', () => {
    describe('#mounted', () => {
      beforeAll(() => {
        wrapper = shallowComponent({
          propsData: {
            name: 'username',
            value: 'initial value'
          }
        })
      })

      it('set in $el "value" prop', () => {
        expect(wrapper.vm.$el.value).toBe('initial value')
      })
    })
  })

  describe('Snapshots', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    it('matches', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    describe('When has "invalid" state', () => {
      beforeAll(() => {
        wrapper.setProps({ state: 'invalid', error: 'username invalid' })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('When has "valid" state', () => {
      beforeAll(() => {
        wrapper.setProps({ state: 'valid' })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('Event Listeners', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    describe('When input changed', () => {
      beforeAll(() => {
        const input = wrapper.find('input')

        jest.spyOn(wrapper.vm, '$emit')
        input.element.value = 'value'
        input.trigger('input')
      })

      it('emits "input" with value', () => {
        expect(wrapper.vm.$emit).toHaveBeenCalledWith('input', 'value')
      })
    })
  })

  describe('Methods', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    describe('#listeners', () => {
      it('has "input" listener', () => {
        expect(wrapper.vm.listeners.input).toBeTruthy()
      })
    })
  })
})
