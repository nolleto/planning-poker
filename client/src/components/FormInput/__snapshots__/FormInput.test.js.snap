// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component Snapshots When has "invalid" state matches 1`] = `
<div>
  <input
    class="formControl isInvalid"
    id="username"
    name="username"
  />
   
  <forminputfeedback-stub
    state="invalid"
  >
    
    username invalid
  
  </forminputfeedback-stub>
</div>
`;

exports[`FormInput component Snapshots When has "valid" state matches 1`] = `
<div>
  <input
    class="formControl isValid"
    id="username"
    name="username"
  />
   
  <forminputfeedback-stub
    state="valid"
  >
    
    username invalid
  
  </forminputfeedback-stub>
</div>
`;

exports[`FormInput component Snapshots matches 1`] = `
<div>
  <input
    class="formControl"
    id="username"
    name="username"
  />
   
  <forminputfeedback-stub>
    
    
  
  </forminputfeedback-stub>
</div>
`;
